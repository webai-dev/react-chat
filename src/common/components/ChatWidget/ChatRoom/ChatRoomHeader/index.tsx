import React, {useContext, useEffect, useState} from "react";
import { Container } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import PersonAddIcon from "@material-ui/icons/PersonAdd";
import CloseIcon from "@material-ui/icons/Close";
import {ChatRoom, RoomDetail} from "../../../../../types/types";
import { RoomListsContext } from "../../../../hooks/useRoomLists";
import { ChatRoomsContext } from "../../../../hooks/useChatRooms";
import { CompanyContext } from "../../../../hooks/useCompanies";
import {MeContext} from "../../../../hooks/useMe";
import { getRoomDetail } from "../../../../services/rooms";
import ChatImage from "../../_Common/ChatImage";

// Global styles
export const useStyles = makeStyles({
  root: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    width: "100%",
    height: "70px",
    padding: "10px 15px",
  },
  closeIcon: {
    marginLeft: "20px",
  },
  detailsContainer: {
    display: "flex",
    flexDirection: "row",
    alignItems: "center",
  },
  iconContainer: {
    cursor: "pointer",
    color: (exp) => (!exp ? "#A4AFB7" : "#FFFFFF"),
  },
  imageContainer: {
    position: "relative",
    width: 50,
    height: 50,
    minWidth: 50,
    marginRight: "20px",
  },
  info: {
    marginLeft: 0,
  },
  name: {
    color: (exp) => (!exp ? "#555555" : "#FFFFFF"),
    fontSize: 16,
    fontWeight: "bold",
  },
  department: {
    color: (exp) => (!exp ? "#555555" : "#FFFFFF"),
    fontSize: 12,
  },
  photo: {
    width: 50,
    height: 50,
    borderRadius: 25,
  },
  status: {
    position: "absolute",
    width: 12,
    height: 12,
    bodorRadius: 8,
    boxShadow: "0px 1px 3px #00000029",
    border: "1px solid #707070",
    borderRadius: 6,
    bottom: 0,
    right: 0,
  },
});

interface ChatRoomHeaderProps {
  chatRoom: ChatRoom | undefined;
  connectionStatus: number;
}

/**
 * ChatRoomHeader component
 * @param {ChatRoomHeaderProps} props Properties
 * @description ChatRoomHeader component consists of User Details, Add Button, Close button
 */
const ChatRoomHeader: React.FC<ChatRoomHeaderProps> = ({
  chatRoom = null,
  connectionStatus = 1,
}) => {
  // Default styling
  const { companyRoomLists, loadingList } = useContext(RoomListsContext);
  const { hideChatRoom, updateChatRooms } = useContext(ChatRoomsContext);
  const { selectedCompany } = useContext(CompanyContext);
  const [roomDetail, setRoomDetail] = useState<RoomDetail | undefined>(undefined);
  const { me } = useContext(MeContext);

  const [showGroupChatIcon, setShowGroupChatIcon] = useState(false);

  useEffect(() => {
    const roomItem =
        companyRoomLists &&
        companyRoomLists
            .map((roomList) => roomList.rooms)
            .flat()
            .find((r) => r && r.id === chatRoom?.roomId);
    if (chatRoom && !loadingList) {
      if (roomItem) {
        setRoomDetail({
          id: roomItem.id,
          imageUrl: roomItem.imageUrl,
          lastMessage: roomItem.lastMessage,
          name: roomItem.name,
          participants: roomItem.lastActiveUsers,
          readBy: [],
          roomType: roomItem.roomType,
          unreadCount: roomItem.unreadCount,
          creator: null,
          teamID: null,
        });
        if(roomItem.roomType !== 0) {
          setShowGroupChatIcon(true);
        }
      } else {
        getRoomDetail(chatRoom.companyId, chatRoom.roomId).then((data) => {
          if (data) {
            setRoomDetail(data);
            if (data.teamID){
              setShowGroupChatIcon(false);
            } else setShowGroupChatIcon(data?.creator?.id === me?.id);
          }
        });
      }
    }
  }, [loadingList]);

  const classes = useStyles(
    (!chatRoom?.expanded || !chatRoom.focused || !chatRoom.scrollAtBottom) && roomDetail && roomDetail.unreadCount > 0
  );

  let roomTitle = "";
  let roomSubTitle = "";

  if (roomDetail) {
    const { name, roomType, participants } = roomDetail;
    const filteredParticipants = roomDetail?.participants.filter(p => me && p.id !== me.id);

    if (roomType === 0) {
      roomTitle = name || "Group chat room";
      roomSubTitle = filteredParticipants.map((user) => user.firstName).join(", ");
    } else {
      roomTitle = filteredParticipants
          ? filteredParticipants[0]?.firstName + " " + filteredParticipants[0]?.lastName
          : "";
      roomSubTitle = filteredParticipants
          ? (filteredParticipants[0]?.companyMembership || [])
              .map((membership) => membership.jobTitle)
              .join(", ")
          : "";
    }
  }

  return (
    <div data-testid="chat-room-header">
      <Container className={classes.root}>
        <div className={classes.detailsContainer}>
          <div className={classes.imageContainer}>
            {roomDetail && (
              <ChatImage
                  connectionStatus={connectionStatus}
                  room={{
                    id: roomDetail.id,
                    lastActiveUsers: roomDetail?.participants.filter(p => me && p.id !== me.id),
                    lastReadMessageID: 0,
                    name: roomDetail.name,
                    roomType: roomDetail?.roomType,
                    unreadCount: roomDetail?.unreadCount,
                  }}
              />
            )}
          </div>
          <div className={classes.info}>
            <div className={classes.name}>{roomTitle}</div>
            <div className={classes.department}>{roomSubTitle}</div>
          </div>
        </div>

        <div className={classes.iconContainer}>
          {
            showGroupChatIcon && (
              <PersonAddIcon
                onClick={(e) => {
                  e.preventDefault();
                  e.stopPropagation();

                  if (chatRoom) {
                    chatRoom.groupModalOpened = true;
                    updateChatRooms({...chatRoom});
                  }
                }}
              />)
          }
          <CloseIcon
            className={classes.closeIcon}
            onClick={(e) => {
              e.preventDefault();
              e.stopPropagation();

              if (roomDetail && selectedCompany) {
                hideChatRoom(roomDetail.id, selectedCompany.id);
              }
            }}
          />
        </div>
      </Container>
    </div>
  );
};

export default ChatRoomHeader;
