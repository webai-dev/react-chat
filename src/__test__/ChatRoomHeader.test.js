import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import ChatRoomHeader from "../common/components/ChatWidget/ChatRoom/ChatRoomHeader";
import renderer from "react-test-renderer";

it("renders without crashing", () => {
  const { getByTestId } = render(<ChatRoomHeader />);
  const chatRoomHeader = getByTestId("chat-room-header");
  expect(chatRoomHeader).toBeTruthy();
});

it("matches snapshot", () => {
  let setShowModal = () => undefined;
  const tree = renderer
    .create(<ChatRoomHeader setShowModal={setShowModal} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
