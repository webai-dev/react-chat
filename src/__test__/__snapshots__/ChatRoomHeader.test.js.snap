// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  data-testid="chat-room-header"
>
  <div
    className="MuiContainer-root makeStyles-root-14 MuiContainer-maxWidthLg"
  >
    <div
      className="makeStyles-detailsContainer-16"
    >
      <div
        className="makeStyles-imageContainer-18"
      />
      <div
        className="makeStyles-info-19"
      >
        <div
          className="makeStyles-name-20 makeStyles-name-25"
        >
          
        </div>
        <div
          className="makeStyles-department-21 makeStyles-department-26"
        >
          
        </div>
      </div>
    </div>
    <div
      className="makeStyles-iconContainer-17 makeStyles-iconContainer-24"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M15 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm-9-2V7H4v3H1v2h3v3h2v-3h3v-2H6zm9 4c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
        />
      </svg>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root makeStyles-closeIcon-15"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
        />
      </svg>
    </div>
  </div>
</div>
`;
